File: Tree/129.java
package Tree;

// 129. Sum Root to Leaf Numbers
// 12-3
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int sumNumbers(TreeNode root) {
        output = 0;
        sum(root, Integer.toString(root.val));
        return output;
    }

    public void sum(TreeNode root, String currentSum){
        
        if (root.left == null && root.right == null){
            output += Integer.parseInt(currentSum);
            return;
        }
    
        if (root.left != null) sum(root.left, concat(currentSum, root.left.val));
        if (root.right != null) sum(root.right, concat(currentSum, root.right.val));
    }
    
    public String concat(String currentSum, int rootVal){
        String b = Integer.toString(rootVal);
        return currentSum+b;
    }

    private int output;

}

File: Tree/129.py
# 129. Sum Root to Leaf Numbers
# 12-4

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def sumNumbers(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        output = self.sum(root, 0, 0)
        return output
    
    def sum(self, root, currentSum, output):
        if root:
            currentSum = currentSum * 10 + root.val
        else:
            return output
        if not root.left and not root.right:
            output += currentSum
        else:
            output = self.sum(root.left, currentSum, output)
            output = self.sum(root.right, currentSum, output)
        return output